### PR TITLE
feat(subscription): let admins set expiry, renew or replace on grant, and grant in bulk

### PR DESCRIPTION
**Pull request by YiKongk**
# ⚠️ 提交说明 / PR Notice
> [!IMPORTANT]
>
> - 请提供**人工撰写**的简洁摘要，避免直接粘贴未经整理的 AI 输出。

## 📝 变更描述 / Description

管理员为用户开通订阅目前仅支持「按套餐默认周期新建一条记录」，无法满足以下常见运营场景：

1. **自定义到期时间**：赠送指定时长或补偿至某一具体日期时，无法指定到期时间；
2. **续期已有订阅**：用户订阅临近到期需要延长时，只能新插一条记录，导致两条 active 记录并存；
3. **批量发放**：活动补偿需面向多个用户时，只能逐个操作。

### 具体改动

**授予模式（`mode`）**：`create` / `renew` / `replace`，默认 `create`；不传该参数时行为与现状完全一致。

- `renew`：定位该用户同套餐生效中的订阅，直接 UPDATE 其 `end_time` 而非新增记录。由于不新增记录，续期不占用 `MaxPurchasePerUser` 名额；无生效记录时自动回退为 `create`。
- `replace`：先将该用户同套餐生效中的记录置为 `cancelled`，再新建一条。两个必要的配套处理：
  - 限购计数不含状态过滤（既有语义，本 PR 未改动），刚被取消的记录仍会计入。因此 replace 仅在**确实取消了记录**的前提下跳过本次限购检查——一取消一新建对生效订阅数为净零，不构成新购买；未取消任何记录时（如用户订阅早已过期）限购照常执行，与 `create` 一致。
  - 新记录创建时用户仍处于套餐的 `upgrade_group`，`PrevUserGroup` 快照将取到空值；若不处理，替换产生的订阅到期后用户将永久保留升级分组（`downgrade_group` 为空、依赖"回退到购买前分组"配置的套餐均受影响）。因此 replace 将被取消记录的 `PrevUserGroup` 带入新记录，保证到期降组链路完整。

**自定义到期时间（`end_time`）**：Unix 秒；传 0 或不传时按套餐周期计算。传入正数时直接将到期时间设为该绝对时间戳（可能早于原到期时间，即缩短）；传 0 时在 `max(现有到期时间, 当前时间)` 基础上叠加一个套餐周期，提前续期不损失剩余时长。UI 对两种语义均有明确提示。

关于配额重置计划：`calcNextResetTime` 在下次重置时间晚于到期时间时返回 0（即关闭重置计划），因此 renew 延长 `end_time` 后需考虑重置计划的恢复；同时 custom 周期的重置基于基准时间，若在周期中途无条件重算，会造成已排定的重置被跳过、后一次又提前触发。最终实现为：**已排定且仍然有效的重置时间保持不变**，仅在其已被清零或被新的到期时间置为无效时重算。两种情形均有测试覆盖。

**批量授予**：新增 `POST /api/subscription/admin/bind/batch`。每个用户使用独立事务，单个用户失败（限购已满、账号不存在等）不回滚其他用户；返回 `success_count` / `failed_count` 及逐用户失败原因，并对入参去重。`failed` 字段显式初始化为空切片，全部成功时序列化为 `[]` 而非 `null`，前端可直接遍历。

**日志**：成功的授予按 #6325 引入的模式为每个用户写入 manage 日志（`subscription.granted` op 描述符，前端按查看者语言渲染）；操作者审计新增 `subscription.admin_grant_batch`，单用户授予的审计参数补充 `mode`。

### 附带修复

`GetDBTimestamp()` 通过全局 `DB` 句柄执行查询，在 `DB.Transaction` 内调用时会额外占用一条连接池连接直至事务结束。订阅相关事务路径改为通过所在事务读取（`getDBTimestampFrom(tx)`）。这同时解决了订阅事务路径无法编写单元测试的问题：测试使用连接数限制为 1 的内存 SQLite，原实现会在等待第二条连接时死锁。

### 前端

- `web/default`：单用户弹窗新增模式选择与 `DateTimePicker`（复用兑换码抽屉的「快捷时长按钮 + 时间选择器」组合）；用户表格批量操作栏新增「批量分配订阅」入口，复用现有 row selection；新增批量结果弹窗，展示成功数与失败明细。到期时间提示文案抽取为共享的 `getEndTimeHint`，renew 模式留空时明确提示「在当前到期时间基础上顺延一个套餐周期」。
- `web/classic`：单用户弹窗新增 `RadioGroup` 模式选择与 `DatePicker`。批量入口未实现——classic 的 `CardTable` 尚无 rowSelection 支持，实现成本高于收益。

两套前端的 i18n 均已补齐全部语言。

> 说明：本 PR 代码实现为 AI 辅助完成（AI-assisted），提交者已逐项审阅并在本地验证。

## 🚀 变更类型 / Type of change
- [ ] 🐛 Bug 修复 (Bug fix) - *请关联对应 Issue，避免将设计取舍、理解偏差或预期不一致直接归类为 bug*
- [x] ✨ 新功能 (New feature) - *重大特性建议先通过 Issue 沟通*
- [ ] ⚡ 性能优化 / 重构 (Refactor)
- [ ] 📝 文档更新 (Documentation)

## 🔗 关联任务 / Related Issue
- 依赖 #6325（本分支包含其 commit；#6325 合并后本 PR 仅余 feature commit）

## ✅ 提交前检查项 / Checklist
- [x] **人工确认:** 我已亲自整理并撰写此描述，没有直接粘贴未经处理的 AI 输出。
- [x] **非重复提交:** 我已搜索现有的 [Issues](https://github.com/QuantumNous/new-api/issues) 与 [PRs](https://github.com/QuantumNous/new-api/pulls)，确认不是重复提交。
- [x] **Bug fix 说明:** 若此 PR 标记为 `Bug fix`，我已提交或关联对应 Issue，且不会将设计取舍、预期不一致或理解偏差直接归类为 bug。
- [x] **变更理解:** 我已理解这些更改的工作原理及可能影响。
- [x] **范围聚焦:** 本 PR 未包含任何与当前任务无关的代码改动。
- [x] **本地验证:** 已在本地运行并通过测试或手动验证，维护者可以据此复核结果。
- [x] **安全合规:** 代码中无敏感凭据，且符合项目代码规范。

## 📸 运行证明 / Proof of Work

后端：

```
go build ./model/ ./controller/ ./router/   → 通过
go vet  ./model/ ./controller/ ./router/    → 通过
go test ./model/ ./controller/ -count=1     → 全部通过

新增测试：
  TestAdminBindSubscriptionRenewExtendsWithoutConsumingLimit
  TestAdminBindSubscriptionRenewFallsBackToCreateWhenNoActive
  TestAdminBindSubscriptionRenewKeepsCustomResetSchedule
  TestAdminBindSubscriptionReplaceCancelsExisting
  TestAdminBindSubscriptionReplaceSucceedsAtPurchaseLimit
  TestAdminBindSubscriptionReplacePreservesPrevUserGroup（含到期后经 ExpireDueSubscriptions 正确降组）
  TestAdminBindSubscriptionHonoursCustomEndTime
  TestAdminBindSubscriptionBatchReportsPerUserFailures
  TestAdminBindSubscriptionBatchAllSuccessKeepsFailedNonNil
```

前端：

```
web/default: bun run typecheck → 通过；bun run build → 通过
web/classic: bun run build     → 通过
```

（三种模式的操作界面与批量分配结果面板截图见下方评论区补充。）

## Summary by CodeRabbit

* **New Features**
  * Added subscription grant modes (create/renew/replace) with optional custom expiration times.
  * Added admin batch subscription assignment with per-user success/failure reporting.
  * Updated subscription grant/expiry audit messaging with new localized templates.
  * Added admin UI for grant mode selection and custom expiration (including preset durations).

* **Bug Fixes**
  * Improved renewal/replacement behavior for active subscriptions, including correct end-time handling and purchase-limit enforcement.
  * Prevented orphan subscription records for invalid users and validated custom end times.

* **Tests**
  * Added database-backed coverage for locking, renew/replace semantics, batch results, and edge cases.

### CODE REVIEW

File: web/default/src/features/subscriptions/components/dialogs/user-subscriptions-dialog.tsx [hunk 1]
```diff
@@ -242,41 +265,115 @@ export function UserSubscriptionsDialog(props: Props) {
...
+                    size='sm'
+                    onClick={() => setEndTime(addTimeToDate(0, 1, 0))}
+                  >
+                    {t('1 Day')}
+                  </Button>
+                </div>
+                <p className='text-muted-foreground text-sm'>
+                  {getEndTimeHint(t, mode, !!endTime)}
+                </p>
+              </div>
```
[the remark this answers is not in the document]
Comment (YiKongk) [reply web/default/src/features/subscriptions/components/dialogs/user-subscriptions-dialog.tsx:376]: This is a false positive. `addTimeToDate(0, 0, 0)` does not return "now" — it explicitly returns `undefined` when all three offsets are zero (`web/default/src/lib/time.ts`):

```ts
export function addTimeToDate(
  months: number,
  days: number,
  hours: number,
  baseDate: Date = new Date()
): Date | undefined {
  if (months === 0 && days === 0 && hours === 0) {
    return undefined
  }
  ...
}
```

So the "Plan default" button performs `setEndTime(undefined)`, which clears the custom expiration; `handleSubmit`/`handleCreate` then send `end_time: 0` (`endTime ? Math.floor(endTime.getTime() / 1000) : 0`), and the backend applies the plan default duration. The same applies to `batch-assign-subscription-dialog.tsx`. No change needed.

**Comment by YiKongk** [comment]
@coderabbitai The remaining findings from the first review pass have been addressed in fd072dc:

- **Effective grant mode in audit logs** (outside-diff, `model/subscription.go`): `AdminBindSubscription` now returns the mode that actually ran; the renew-with-no-active fallback reports `create`, and `recordSubscriptionGrantLogs` records the effective mode. Covered by an assertion in `TestAdminBindSubscriptionRenewFallsBackToCreateWhenNoActive`.
- **`NormalizeGrantMode` strictness**: unknown non-empty modes are now rejected with an error instead of silently coercing to `create`; empty input keeps the historical default. Batch validates once up front. Covered by a new negative test.
- **Redundant group query**: `createUserSubscriptionFromPlanTx` reuses `lockedUser.Group` instead of re-querying via `getUserGroupByIdTx`.
- **Batch dedup test**: `TestAdminBindSubscriptionBatchReportsPerUserFailures` now feeds duplicated ids (`[101, 101, 102, 102, 999]`) and asserts unique processing plus a single row for the succeeded user.
- **`mode` typing in `constants.ts`**: `getGrantModeDescription` / `getEndTimeHint` now take `SubscriptionGrantMode`.

Deferred to a follow-up PR to keep this one focused:
- Extracting the shared grant-and-respond controller helper.
- Extracting a shared `GrantExpirationFields` component for the two dialogs.
